src/client/datascience/notebook/notebookControllerManager.ts

```diff
@@ -131,6 +137,42 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
         return this.controllersPromise;
     }
 
+    public getOrCreateController(
+        pythonInterpreter: PythonEnvironment,
+        notebookType: 'interactive' | 'jupyter-notebook'
+    ): VSCodeNotebookController | undefined {
+        // If already registered just return it
+        let matchingController: VSCodeNotebookController | undefined;
+        if (this.registeredControllers.size > 0) {
+            matchingController = this.registeredNotebookControllers().find((controller) => {
+                return (
+                    // We register each of our kernels as two controllers
+                    // because controllers are currently per-viewtype. Find
+                    // the one for the interactive viewtype for now
+                    controller.controller.notebookType === notebookType &&
+                    controller.connection.kind === 'startUsingPythonInterpreter' &&
+                    controller.connection.interpreter?.path === pythonInterpreter?.path &&
+                    controller.connection.interpreter.displayName === pythonInterpreter.displayName
+                );
+            });
+        }
+        if (!matchingController) {
+            // Otherwise try to create the VSCodeNotebookController
+            const spec = createInterpreterKernelSpec(pythonInterpreter);
+            const result: PythonKernelConnectionMetadata = {
+                kind: 'startUsingPythonInterpreter',
+                kernelSpec: spec,
+                interpreter: pythonInterpreter,
+                id: getKernelId(spec, pythonInterpreter)
+            };
+            const matchingControllers = this.createNotebookControllers([result]);
+            matchingController = matchingControllers.find(
```

Comment:
Can we not iterate `this.registeredControllers` to find the new item, rather than changing the return value in two places. Feels a little easier to follow the code.